### ISSUE TITLE
ascii csv file: Header/data length mismatch

### ISSUE THREAD

**Issue by jungle-boogie** [johnkerl/miller#103]
Hi John,

I'd like to use mlr to process some trip data with uber.

Here's part of the file with `cat -e`:

```
trip_id,date,dayOfWeek,pickup_time,duration,distance,surge,surge_amount,total,beginning_zip,ending_zip,tip,amount,sex,number_of_riders$
f,2016-04-18,monday,18:42,1036.8,3.17,no,,6.7,89119,89109$
1,2016-04-18,monday,19:05,816,3.35,no,,6.31,89109,89109$
6,2016-04-18,monday,19:44,970.8,5.52,no,,8.48,89109,89119$
c,2016-04-18,monday,20:28,871.8,3.8,no,,6.82,89123,89119$
e,2016-04-18,monday,21:22,768,7.84,no,,9.88,89183,89109$
```

`% file uber.csv`
`uber.csv: ASCII text`

`% mlr --rs lf --csv  cat uber.csv`
`mlr: Header/data length mismatch (15 != 11) at file "uber.csv" line 2.`

I think this may be because not all of my records have:
`surge`
`surge_amount`
`tip`
`amount`
`sex`
`number_of_riders`

any hints on how I can work with this file while preserving the header names so I can later reference them?

**Comment by johnkerl** [comment]
This is definitely broken CSV -- the header line has 15 fields while the data lines have 11. You can see this more clearly here:

```
 head -n 3 jb.csv | mlr --inidx --ifs , --oxtab cat
1  trip_id
2  date
3  dayOfWeek
4  pickup_time
5  duration
6  distance
7  surge
8  surge_amount
9  total
10 beginning_zip
11 ending_zip
12 tip
13 amount
14 sex
15 number_of_riders

1  f
2  2016-04-18
3  monday
4  18:42
5  1036.8
6  3.17
7  no
8  
9  6.7
10 89119
11 89109

1  1
2  2016-04-18
3  monday
4  19:05
5  816
6  3.35
7  no
8  
9  6.31
10 89109
11 89109
```
Since the fields `tip,amount,sex,number_of_riders` are the ones missing, you might pre-process with `sed` to fix your broken CSV. For example:

```
 sed 's/,tip.*//' x | mlr --icsvlite --opprint cat
trip_id date       dayOfWeek pickup_time duration distance surge surge_amount total beginning_zip ending_zip
f       2016-04-18 monday    18:42       1036.8   3.17     no    -            6.7   89119         89109
1       2016-04-18 monday    19:05       816      3.35     no    -            6.31  89109         89109
6       2016-04-18 monday    19:44       970.8    5.52     no    -            8.48  89109         89119
c       2016-04-18 monday    20:28       871.8    3.8      no    -            6.82  89123         89119
e       2016-04-18 monday    21:22       768      7.84     no    -            9.88  89183         89109
```

**Comment by johnkerl** [comment]
You can also push back on whoever's generating this data and calling it "CSV".

**Comment by jungle-boogie** [comment]
So regarding field 8, just fill it in with 0?

**Comment by johnkerl** [comment]
If you like -- null and 0 are different things. Depends on what you want to do with the data.

**Comment by jungle-boogie** [comment]
Thanks, John. 

Sadly not all data is known for the latter columns but mlr doesn't care about those. I may leave them blank for the time being.

**Comment by johnkerl** [comment]
Ohh if you're creating the data -- then the standard thing people do to their CSV files for data which is not known / applicable for all fields is to leave them empty, but still put the commas in. In this particular data that would amount to four trailing commas on the data lines:

```
$ cat jb.csv
trip_id,date,dayOfWeek,pickup_time,duration,distance,surge,surge_amount,total,beginning_zip,ending_zip,tip,amount,sex,number_of_riders
f,2016-04-18,monday,18:42,1036.8,3.17,no,,6.7,89119,89109,,,,
1,2016-04-18,monday,19:05,816,3.35,no,,6.31,89109,89109,,,,
6,2016-04-18,monday,19:44,970.8,5.52,no,,8.48,89109,89119,,,,
c,2016-04-18,monday,20:28,871.8,3.8,no,,6.82,89123,89119,,,,
e,2016-04-18,monday,21:22,768,7.84,no,,9.88,89183,89109,,,,
```

```
$ mlr --icsvlite --oxtab cat jb.csv
trip_id          f
date             2016-04-18
dayOfWeek        monday
pickup_time      18:42
duration         1036.8
distance         3.17
surge            no
surge_amount     
total            6.7
beginning_zip    89119
ending_zip       89109
tip              
amount           
sex              
number_of_riders 

trip_id          1
date             2016-04-18
dayOfWeek        monday
pickup_time      19:05
duration         816
distance         3.35
surge            no
surge_amount
total            6.31
beginning_zip    89109
ending_zip       89109
tip              
amount           
sex              
number_of_riders 

...
```

**Comment by johnkerl** [comment]
In my general experience it's better to get the header length == data length correct when generating the data, which is what I recommended above.

But to fix it up after the fact (which is what I now realize I think was your original question) ... you could use `awk` for sure. Also you could use Miller. This requires the looping feature which will be in the 4.1.0 release (not out yet) but it's available if you pull from head.

The trick is to temporarily treat the file as NIDX format (like `awk`) which doesn't insist that each line have the same number of fields. Then remember the max length encountered so far, and pad out extra fields to reach that maximum field count. For example:

```
mlr --fs comma --nidx put '
  @maxnf = max(@maxnf, NF);
  @nf = NF;
  while(@nf < @maxnf) {
    @nf += 1;
    $[@nf] = ""
  }
' jb.csv
```

**Comment by johnkerl** [comment]
However, NIDX format isn't aware of the fundamental CSV double-quoting feature which allows you to put commas inside of field names, so this solution can't be used to patch up arbitrarily broken CSV files.

**Comment by jungle-boogie** [comment]
Hi John,

What I did with surge_amount was fill in all blank entries with `null`. Now that I see you didn't need to do that, I deleted null, left the additional comma and it works now as expected. I think `mlr` had issues with the surge_amount field over any other field.

Looking at the original data, I do see my `cat -e` example didn't contain the additional commas on the lattter columns and right, that's definitely a broken csv.

I tried out the NIDX format and it worked! In my case now that I have a good csv file, it's not necessary, but I do like the NIDX addition to mlr. 

So overall, this whole thing was my mistake and mlr treated the bad formatted csv file as it should have.
